include/boost/random/niederreiter_base2.hpp

```diff
@@ -0,0 +1,371 @@
+        term ^= v[r + k];
+    }
+    v[r] = term;
+  }
+}
+
+} // namespace nb2
+
+template<typename IntType, typename Nb2Table>
+struct niederreiter_base2_lattice
+{
+  typedef IntType value_type;
+
+  BOOST_STATIC_CONSTANT(unsigned, bit_count = std::numeric_limits<value_type>::digits);
+
+  explicit niederreiter_base2_lattice(std::size_t dimension)
+  {
+    resize(dimension);
+  }
+
+  void resize(std::size_t dimension)
+  {
+    detail::dimension_assert("Niederreiter base 2", dimension, Nb2Table::max_dimension);
+
+    // Initialize the bit array
+    bits.resize(boost::extents[bit_count][dimension]);
+
+    // Reserve temporary space for lattice computation
+    boost::multi_array<value_type, 2> ci(boost::extents[bit_count][bit_count]);
+
+    std::vector<value_type> v;
+
+    // Compute Niedderreiter base 2 lattice
+    for (std::size_t dim = 0; dim != dimension; ++dim)
+    {
+      const unsigned int poly = Nb2Table::polynomial(dim);
+      if (poly > std::numeric_limits<value_type>::max()) {
+        boost::throw_exception( std::range_error("niederreiter_base2: polynomial value outside the given value type range") );
+      }
+
+      const unsigned degree = multiprecision::msb(poly); // integer log2(poly)
+      const unsigned max_degree = degree * ((bit_count / degree) + 1);
+
+      v.resize(degree + max_degree);
+
+      // For each dimension, we need to calculate powers of an
+      // appropriate irreducible polynomial, see Niederreiter
+      // page 65, just below equation (19).
+      // Copy the appropriate irreducible polynomial into PX,
+      // and its degree into E.  Set polynomial B = PX ** 0 = 1.
+      // M is the degree of B.  Subsequently B will hold higher
+      // powers of PX.
+      boost::dynamic_bitset<>::size_type pb_degree = 0;
+      boost::dynamic_bitset<> pb(max_degree, 1);
+
+      unsigned j = 0;
+      while (j != bit_count)
+      {
+        // Now multiply B by PX so B becomes PX**J.
+        // In section 2.3, the values of Bi are defined with a minus sign :
+        // don't forget this if you use them later!
+        nb2::modulo2_multiply(poly, boost::move(pb)).swap(pb);
+
+        // If U = 0, we need to set B to the next power of PX
+        // and recalculate V.
+        nb2::calculate_v(pb, pb_degree, v);
+
+        // Niederreiter (page 56, after equation (7), defines two
+        // variables Q and U.  We do not need Q explicitly, but we
+        // do need U.
+
+        // Advance Niederreiter's state variables.
+        for (unsigned u = 0; u != degree && j != bit_count; ++u, ++j)
+        {
+          // Now C is obtained from V.  Niederreiter
+          // obtains A from V (page 65, near the bottom), and then gets
+          // C from A (page 56, equation (7)).  However this can be done
+          // in one step.  Here CI(J,R) corresponds to
+          // Niederreiter's C(I,J,R).
+          for (unsigned r = 0; r != bit_count; ++r) {
+            ci[r][j] = v[r + u];
```

Comment:
I get your idea that you want to put, say ci's column to the function that computes and, perhaps, avoid v altogether, but I don't think that idea will fly. At least not without some major algorithmic complications.

As for the iteration order, I iterate in column major here so that I could iterate in row major a few lines below. ;-) If I change the order in one place, I will need to change the order in another place.